Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler)]() mutable {
+                completionHandler(Exception { SecurityError });
+            });
+            return;
+        }
+
+        auto url = document->url();
+        if (!options.url.isNull()) {
+            auto parsed = document->completeURL(options.url);
+            if (globalObjectIsWindowObject && parsed != url) {
+                ensureOnContextThread([completionHandler = WTFMove(completionHandler)]() mutable {
+                    completionHandler(Exception { TypeError });
+                });
+                return;
+            }
+
+            if (!origin->isSameOriginDomain(SecurityOrigin::create(parsed))) {
+                ensureOnContextThread([completionHandler = WTFMove(completionHandler)]() mutable {
+                    completionHandler(Exception { TypeError });
+                });
+                return;
+            }
+            url = WTFMove(parsed);
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler), cookies = WTFMove(cookies)]() mutable {
+                if (!cookies)
+                    completionHandler(Exception { TypeError });
+                else
+                    completionHandler(WTFMove(*cookies));
+            });
+        };
+
+        cookieJar.getCookiesAsync(document, url, options, WTFMove(resultHandler));
+    };
+
+    ensureOnMainThread(WTFMove(getAllCookies));
+}
+
+void CookieStore::MainThreadBridge::set(CookieInit&& options, CompletionHandler<void(ExceptionOr<void>&&)>&& completionHandler)
+{
+    auto setCookie = [this, protectedThis = Ref { *this }, options = crossThreadCopy(WTFMove(options)), completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler)]() mutable {
+                completionHandler(Exception { SecurityError });
+            });
+            return;
+        }
+
+        // The maximum attribute value size is specified at https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size.
+        static constexpr auto maximumAttributeValueSize = 1024;
+
+        Cookie cookie;
+        cookie.name = WTFMove(options.name);
+        cookie.value = WTFMove(options.value);
+        cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
+
+        cookie.domain = options.domain.isNull() ? document->domain() : WTFMove(options.domain);
+        if (!cookie.domain.isNull()) {
```

Comment:
Seems like most or all of these checks don't need to happen on the main thread and could be moved to CookieStore instead of the MainThreadBridge.

Reply:
But these checks require us to have a document. If we do them on the caller side, where will we get the document in the case where the caller side is called on a service worker thread?

Reply:
They don't really require a Document, a ScriptExecutionContext should suffice.

From what I gather, you need the domain and the URL host. The domain is `securityOrigin().domain()` which you can call on ScriptExecutionContext. The host can be retrieved from the securityOrigin() too.